python/cuvs/cuvs/neighbors/prefilters/prefilters.pyx

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# cython: language_level=3
+
+import numpy as np
+
+from libc.stdint cimport uintptr_t
+
+from cuvs.common cimport cydlpack
+
+from .prefilters cimport BITMAP, NO_FILTER, cuvsPrefilter
+
+from pylibraft.common import auto_convert_output, cai_wrapper, device_ndarray
+from pylibraft.common.cai_wrapper import wrap_array
+from pylibraft.neighbors.common import _check_input_array
+
+
+@auto_convert_output
+def no_filter():
+    """
+    Create a default pre-filter which filters nothing.
+    """
+    cdef cuvsPrefilter filter = cuvsPrefilter()
+    filter.type = NO_FILTER
+    filter.addr = <uintptr_t>NULL
+
+    return filter
+
+
+@auto_convert_output
+def from_bitmap(bitmap):
+    """
+    Create a pre-filter from an array with type of uint32.
+
+    Parameters
+    ----------
+    bitmap : numpy.ndarray
+        An array with type of `uint32` where each bit in the array corresponds
+        to if a sample and query pair is greenlit (not filtered) or filtered.
+        The array is row-major, meaning the bits are ordered by rows first.
+        Each bit in a `uint32` element represents a different sample-query
+        pair.
+
+        - Bit value of 1: The sample-query pair is greenlit (allowed).
+        - Bit value of 0: The sample-query pair is filtered.
+
+    Returns
+    -------
+    index : cuvs.neighbors.cuvsPrefilter
+        An instance of `cuvsPrefilter` that can be used to filter neighbors
+        based on the given bitmap.
+    {resources_docstring}
+
+    Examples
+    --------
+
+    >>> import cupy as cp
+    >>> import numpy as np
+    >>> from cuvs.neighbors import prefilters
+    >>>
+    >>> n_samples = 50000
+    >>> n_queries = 1000
+    >>>
+    >>> n_bitmap = np.ceil(n_samples * n_queries / 32).astype(int)
+    >>> bitmap_arr = cp.random.random_sample((n_bitmap), dtype=cp.uint32)
+    >>> prefilter = prefilters.from_bitmap(bitmap_arr)
+    """
+    bitmap_cai = wrap_array(bitmap)
+    _check_input_array(bitmap_cai, [np.dtype('uint32')])
+
+    cdef cydlpack.DLManagedTensor* bitmap_dlpack = \
+        cydlpack.dlpack_c(bitmap_cai)
+
+    cdef cuvsPrefilter filter = cuvsPrefilter()
+    filter.type = BITMAP
+    filter.addr = <uintptr_t>bitmap_dlpack
```

Comment:
How memory safe is this? If we go something like

```
prefilter = prefilters.from_bitmap(bitmap)
bitmap = None

```
does the `prefilter.addr` still point to valid memory, or has it been freed up when the `bitmap` refcnt goes down to 0?

(this might not be an issue, but I'm just curious - since most of the time the dlpack wrappers we create are temporaries and the wrapped objects won't be freed up, so we can't get dangling pointers.).

Reply:
I'm also curious on it, I suppose the `cupy` should manage the HBM like python does on host memory, assigning a `none` would not invalidate the bottom address directly. (`brute_force.build` should encounter a similar issue for  `dataset` being set to `None` by a naughty user :P). But for careful consideration, we should seek an answer from a `cupy` team member and the author of `cydlpack.DLManagedTensor`? Hi @cjnolet, could you help recommend an expert on it?

Reply:
This is more of a python question than a cupy question -  I think the cydlpack code doesn't take ownership of an array, it just provides an interface so other libraries can consume that array. If the memory in the array gets cleaned up - I think this dlpack wrapper will still point to the previous memory address 

I think we could get around this pretty easily if this is a problem - we just need to hold on to a python reference to the passed in object .

Reply:
I just took a quick look at this - and can verify that this is a problem. Changing the unittest to free the bitmap_device after creating the filter fails - with either cuda memory errors or invalid results:

```
diff --git a/python/cuvs/cuvs/test/test_brute_force.py b/python/cuvs/cuvs/test/test_brute_force.py
index caa5d65..526a51d 100644
--- a/python/cuvs/cuvs/test/test_brute_force.py
+++ b/python/cuvs/cuvs/test/test_brute_force.py
@@ -145,6 +145,7 @@ def test_prefiltered_brute_force_knn(
     distances_device = device_ndarray(distances)
     bitmap_device = device_ndarray(bitmap)
     prefilter = prefilters.from_bitmap(bitmap_device)
+    bitmap_device = None
 
     brute_force_index = brute_force.build(index_device, metric)
     ret_distances, ret_indices = brute_force.search(
```

I don't think this is really something we need to fix in the DLManagedTensor class : DLManagedTensor is just a thin wrapper around dlpack, and already supports things like moving ownership. The issue is our use of the class here - we're attempting to take a non-owning view of the underlying memory in the `prefilters.from_bitmap` code.

It should be straightforward to adjust the prefilter api to keep a reference to the memory, so that this isn't a problem. This will involve creating a wrapper python object for `cuvsPrefilter` (rather than just having the python API use the C `cuvsPrefilter` object directly). This wrapper will have a python member to keep the refcnt for the underlying cuda memory alive. 

As an example of what I'm suggesting, take a look at the cagra 'IndexParams' object https://github.com/rapidsai/cuvs/blob/9dc3a4dc5085cb5eb92d18ca1eafd5eef3dc405e/python/cuvs/cuvs/neighbors/cagra/cagra.pyx#L153-L154 . This IndexParams object holds a 'CompressionParams' reference, to avoid similar errors.